staging/src/k8s.io/api/flowcontrol/v1alpha1/types.go

```diff
@@ -0,0 +1,297 @@
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// PriorityLevelConfigurationList is a list of PriorityLevelConfiguration objects.
+type PriorityLevelConfigurationList struct {
+	metav1.TypeMeta `json:",inline"`
+	// `metadata` is the standard object's metadata.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	metav1.ListMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	// `items` is a list of request-priorities.
+	Items []PriorityLevelConfiguration `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+// PriorityLevelConfigurationSpec is specification of a priority level
+type PriorityLevelConfigurationSpec struct {
+	// `exempt` defines whether the priority level is exempted or not.  There should be at most one exempt priority level.
+	// Being exempt means that requests of that priority are not subject to concurrency limits (and thus are never queued)
+	// and do not detract from the concurrency available for non-exempt requests. In a more sophisticated system, the
+	// exempt priority level would be the highest priority level. The field is default to false and only those system
+	// preset priority level can be exempt.
+	// +optional
+	Exempt bool `json:"exempt,omitempty" protobuf:"varint,1,opt,name=exempt"`
+	// `assuredConcurrencyShares` is a positive number for a non-exempt priority level, representing the weight by which
+	// the priority level shares the concurrency from the global limit. The concurrency limit of an apiserver is divided
+	// among the non-exempt priority levels in proportion to their assured concurrency shares. Basically this produces
+	// the assured concurrency value (ACV) for each priority level:
+	//
+	//             ACV(l) = ceil( SCL * ACS(l) / ( sum[priority levels k] ACS(k) ) )
+	//
+	// +optional
+	AssuredConcurrencyShares int32 `json:"assuredConcurrencyShares,omitempty" protobuf:"varint,2,opt,name=assuredConcurrencyShares"`
+	// `queues` is a number of queues that belong to a non-exempt PriorityLevelConfiguration object. The queues exist
+	// independently at each apiserver. The value must be positive for a non-exempt priority level and setting it to 1
+	// disables shufflesharding and makes the distinguisher method irrelevant.
+	// +optional
+	Queues int32 `json:"queues,omitempty" protobuf:"varint,3,opt,name=queues"`
+	// `handSize` is a small positive number for applying shuffle sharding. When a request arrives at an apiserver the
+	// request flow identifier’s string pair is hashed and the hash value is used to shuffle the queue indices and deal
+	// a hand of the size specified here. If empty, the hand size will the be set to 1.
+	// +optional
+	HandSize int32 `json:"handSize,omitempty" protobuf:"varint,4,opt,name=handSize"`
+	// `queueLengthLimit` is a length limit applied to each queue belongs to the priority.  The value must be positive
+	// for a non-exempt priority level.
```

Comment:
is there also a limit on time? How do I decide what this limit should be?

Reply:
There is a limit on waiting duration, set automatically to one quarter of the handling timeout (the apiserver parameter that defaults to one minute).  The waiting duration limit is described in the KEP and the doc.  We did not mention it here because this data type is not the source of that limit.

Reply:
Something very approximately like `queue length * (average request processing time for this type of requests with this many shares) ~ wait time` is true of this setup, right?

I think this implies that there's a large range of lengths which are useless-- they never get used because the time in practice always triggers first. I think it might make sense to mention that there is also a time limit--otherwise I can easily imagine an admin making this value super large because they're not aware it's actually the time limit being hit.

Or maybe I misunderstand. In which case it'd be good to say something in the comment which makes me less confused :)

Reply:
The math is a little more complicated than you outlined, and there are other reasons to keep the length limit low.  Even if the apiserver applied no timeout to requests in the queue, the handling timeout filter is earlier in the handler chain so that applies.  And even if that did not apply, nobody wants things waiting in queues for a long time.

I am not sure how much discussion belongs in this data type's comments.  I think we should have a good presentation in the doc and/or blog.

Reply:
i was completely convinced about `wait-time=global-request-timeout/4` yet.. we can probably move this as a data-type in the future.. but before that, i think we need some benchmark experiment on how the wait-time/queue-length affects the system for kube-apiserver

Reply:
Yes, we should get some experience, including changing the automatic derivation if we do not like the first attempt, and expose the ability to control the limit if we get evidence that there is no good enough formula for an automatic setting.

Reply:
OK, I think a brief heads-up and pointer would be OK; I think more would be out of place.